### PR TITLE
Implement `spread` intrinsic

### PR DESCRIPTION
**Pull request by HarshitaKalani**
Fixes: https://github.com/lfortran/lfortran/issues/4408, https://github.com/lfortran/lfortran/issues/2433, https://github.com/lfortran/lfortran/issues/1315
Towards: #4017, #492

### CODE REVIEW

**Comment by HarshitaKalani** [comment]
There can be a better version of this implementation. This works for now, will improve the implementation in subsequent PRs.

**Comment by certik** [comment]
I think the Windows error is real:
```
2024-07-01T07:20:14.9402464Z D:\a\lfortran\lfortran\lfortran-0.37.0-14-g19e16880f\src\libasr\..\libasr/pass/intrinsic_array_function_registry.h(1460): error C2146: syntax error: missing ')' before identifier 'or'
2024-07-01T07:20:15.3451530Z D:\a\lfortran\lfortran\lfortran-0.37.0-14-g19e16880f\src\libasr\..\libasr/pass/intrinsic_array_function_registry.h(1460): error C2143: syntax error: missing ';' before '{'
```

**Comment by certik** [comment]
Awesome, thank you for this!

I think spread creates a 2D array out of 1D. So we should test it in the test that it works this way.

File: src/libasr/pass/intrinsic_array_function_registry.h [hunk 1]
```diff
@@ -1436,6 +1438,316 @@ namespace Cshift {
...
+        ASRUtils::require_impl(x.m_args[2], "`ncopies` argument of `spread` "
+            "cannot be nullptr", x.base.base.loc, diagnostics);
+    }
+
+    static ASR::expr_t *eval_Spread(Allocator &al, const Location &loc,
+            ASR::ttype_t *type, Vec<ASR::expr_t*> &args, diag::Diagnostics& diag) {
+        ASRBuilder b(al, loc);
+        int dim_val = ASR::down_cast<ASR::IntegerConstant_t>(ASRUtils::expr_value(args[1])) -> m_n;
+        if (all_args_evaluated(args) &&
+            (extract_n_dims_from_ttype(expr_type(args[0])) == 1) && (dim_val == 1 or dim_val == 2)) {
```
Comment (certik) [inline src/libasr/pass/intrinsic_array_function_registry.h:1460]: This line gives the error message on Windows:
```
2024-07-02T06:52:30.6302614Z D:\a\lfortran\lfortran\lfortran-0.37.0-31-g4577497bc\src\libasr\..\libasr/pass/intrinsic_array_function_registry.h(1460): error C2146: syntax error: missing ')' before identifier 'or'
2024-07-02T06:52:30.6322973Z D:\a\lfortran\lfortran\lfortran-0.37.0-31-g4577497bc\src\libasr\..\libasr/pass/intrinsic_array_function_registry.h(1460): error C2143: syntax error: missing ';' before '{'
```
And to fix it, you can't use `or`:
```suggestion
            (extract_n_dims_from_ttype(expr_type(args[0])) == 1) && (dim_val == 1 || dim_val == 2)) {
```

**Review by certik — APPROVED** [review]

File: src/libasr/pass/intrinsic_array_function_registry.h (hunk 1 above)
Comment (HarshitaKalani) [reply src/libasr/pass/intrinsic_array_function_registry.h]: Oh I get this. Thank you!

**Comment by HarshitaKalani** [comment]
>I think spread creates a 2D array out of 1D. So we should test it in the test that it works this way.

In order to test it that way, we need reshape function which currently gives junk values and hence I've done it the way as in this PR. I hope works for now.

**Comment by certik** [comment]
Yes, the tests are fine, thanks!

**Pull request merged**